### PR TITLE
[MC-40] Create Component Logic Password Input

### LINKED ISSUE #40: [Feature] Create Component Logic Password Input
**Description**
I want a class to be created that represents the password entry that will be used as the default for the entire site, being possible to instantiate the object for each component.

The following validations must be considered:
- password = same as input + must be at least size 8, contain at least 1 capital letter, 1 number, and 1 special character

**Additional context**
const pwi1 = new PasswordIpunt('btn query 1');
const pwi2 = new PasswordIpunt('btn query 2');

**Acceptance Criteria**
- The component must have its class
- The component needs to keep its state (valid/invalid, disabled/enabled)
- The component must have a get/set value
- The component must have validation
- The component must have required (true|false)
- The component is expected to have tests

### PR DESCRIPTION
**Pull request by pedroemanuel42**
# Feature/Issue Description

This PR create a class password components that has get/set values, validations, required and disabled checks with tests around then.

# Changes Summary

- Add get/set Value;
- Add password input validations (must be same as input, at least size 8, contain at least 1 capital letter, 1 number, and 1 special character);
- Add get/set valid, enabled and disabled;
- Add verify into valid, enabled and disabled;
- Add tests to all this new methods.

# How Has This Been Tested?

To test all methods, i created new obects using them, checked in the localhost log to see these returns.

### CODE REVIEW

**Comment by tvSantiago** [comment]
Your description of how has this been tested is confusing.
First don't possible how to know test. Must create a new object? Or just run the server?

File: src/assets/scss/_variables.scss [hunk 1]
```diff
@@ -28,4 +28,4 @@ $primary-light: #474787;
 $accent-dark: #cd6133;
 $accent-light: #ff793f;
 $other-dark: #d1ccc0;
-$other-light: #ffffff;
\ No newline at end of file
+$other-light: #ffffff;
```
Comment (tvSantiago) [inline src/assets/scss/_variables.scss:31]: Must have an empty line in the file end

File: src/assets/scss/dropdown.scss [hunk 2]
```diff
@@ -87,4 +87,3 @@
     pointer-events: none;
     color: $other-dark;
 }
-
```
Comment (tvSantiago) [inline src/assets/scss/dropdown.scss:90]: Must have an empty line in the file end

File: src/assets/scss/home-layout.scss [hunk 3]
```diff
@@ -35,4 +35,3 @@ $nav-padding: 10px;
     display: flex;
     padding: 10px 10%;
 }
-
```
Comment (tvSantiago) [inline src/assets/scss/home-layout.scss:38]: Must have an empty line in the file end

File: src/assets/scss/input-password.scss [hunk 4]
```diff
@@ -31,18 +30,18 @@
...
 .eye-icon {
+    color: $other-dark;
     display: flex;
     flex-direction: row-reverse;
     position: relative;
-    top: -31px;
-    left: -17px; 
+    top: -32px;
+    left: -17px;
 }
```
Comment (tvSantiago) [inline src/assets/scss/input-password.scss:46]: Must have an empty line in the file end

File: src/components/password.components.js [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+        }
+
+        return this.valid;
+    }
+
+    getDisabled() {
+        return this.input.prop('disabled');
+    }
+
+    setDisabled(boolean) {
```
Comment (tvSantiago) [inline src/components/password.components.js]: This function when passing a value true changes the style:
image
I can also click on the icon to show or hide a password:
image

File: src/components/password.components.js [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+
+    getValue() {
+        return this.input.val();
+    }
+
+    setValue(string) {
+        this.input.val(string);
+    }
+
+    isValid() {
```
Comment (tvSantiago) [inline src/components/password.components.js]: There's a logical problem here
image

File: src/components/password.components.js [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+    getDisabled() {
+        return this.input.prop('disabled');
+    }
+
+    setDisabled(boolean) {
+        this.input.prop('disabled', boolean);
+        this.disabled = boolean;
+    }
+
+    isDisabled() {
```
Comment (tvSantiago) [inline src/components/password.components.js]: This is redundant, this is also for the other properties
image

File: src/components/password.components.js [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+
+        this.valid = valid;
+        this.disabled = disabled;
+        this.required = required;
+
+        this.hasQuery(this.div, this.input, this.icon);
+        this.initFor();
+    }
+
+    hasQuery(div, input, icon) {
```
Comment (tvSantiago) [inline src/components/password.components.js]: These names do not reflect what the functions are doing
image

**Review by tvSantiago — CHANGES_REQUESTED** [review]

**Comment by pedroemanuel42** [comment]
Solved, waiting review again. Thanks for the advices!

**Comment by pedroemanuel42** [comment]
> Your description of how has this been tested is confusing. First don't possible how to know test. Must create a new object? Or just run the server?

Edited, thanks!

File: src/assets/scss/_variables.scss (hunk 1 above)
Comment (tvps20) [reply src/assets/scss/_variables.scss:31]: image

File: src/pages/components/style.scss [hunk 9]
```diff
@@ -59,4 +60,3 @@ p {
     background-color: $primary-light;
     color: $other-light;
 }
```
Comment (tvps20) [inline src/pages/components/style.scss:62]: Must have an empty line in the file end

File: src/pages/components/main.js [hunk 10]
```diff
@@ -1,13 +1,14 @@
...
 import './style.scss';
 import page from './index.html?raw';
 
 LayoutProviderService.inject(page).then(() => {
+    // eslint-disable-next-line no-unused-vars
+    const password = new PasswordInput('#password');
     // eslint-disable-next-line no-unused-vars
     const dropdownGameType = new DropdownComponent('#games');
-    PasswordInputService.initFor('#password');
 });
```
Comment (tvps20) [inline src/pages/components/main.js:13]: Must have an empty line in the file end

File: src/components/tests/password.components.test.js [hunk 11]
```diff
@@ -0,0 +1,254 @@
...
+            `;
+
+            $(document.body).html(div);
+
+            password = new PasswordInput('#pass');
+
+            expect(password.isRequired()).toBeTruthy();
+        });
+    });
+});
```
Comment (tvps20) [inline src/components/tests/password.components.test.js:249]: Must have an empty line in the file end

File: src/components/password.components.js [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+
+    getValue() {
+        return this.input.val();
+    }
+
+    setValue(string) {
+        this.input.val(string);
+    }
+
+    isValid() {
```
Comment (feelsantiago) [inline src/components/password.components.js]: You guys must use the same name for the validation method. And must have the same signature and return.

File: src/components/tests/password.components.test.js [hunk 13]
```diff
@@ -0,0 +1,254 @@
...
+            password = new PasswordInput('#pass');
+            password.setRequired(false);
+
+            const inputProperty = $('#input-password').prop('required');
+
+            expect(inputProperty).toBeFalsy();
+        });
+    });
+
+    describe('[isProperty]', () => {
```
Comment (feelsantiago) [inline src/components/tests/password.components.test.js]: what is isProperty?

**Review by feelsantiago — CHANGES_REQUESTED** [review]

File: src/components/password.components.js [hunk 14]
```diff
@@ -0,0 +1,89 @@
...
+        this.input.prop('required', boolean);
+    }
+
+    isRequired() {
+        if (this.input.prop('required')) {
+            return true;
+        }
+
+        return false;
+    }
```
Comment (tvps20) [inline src/components/password.components.js:89]: The state is not save
image

File: src/components/password.components.js (hunk 7 above)
Comment (tvps20) [reply src/components/password.components.js]: If I click on any point of the component it activates the show/hide should only occur by clicking on the icon
image

File: src/components/password.components.js [hunk 15]
```diff
@@ -0,0 +1,89 @@
...
+        if (input.length === 0) {
+            throw new Error('Failed to query input password');
+        }
+
+        if (icon.length === 0) {
+            throw new Error('Failed to query icon');
+        }
+    }
+
+    onClick() {
```
Comment (tvps20) [inline src/components/password.components.js]: This name does not reflect the comportment

**Review by tvps20 — CHANGES_REQUESTED** [review]

**Review by feelsantiago — DISMISSED** [review]

**Comment by tvps20** [comment]
Pr title is wrong
image

File: src/components/password.components.js [hunk 16]
```diff
@@ -0,0 +1,97 @@
...
+
+export default class PasswordInput {
+    constructor(queryString, valid = false, disabled = false, required = true) {
+        this.div = $(queryString);
+        this.input = $(`${queryString} input[type="password"]`);
+        this.icon = $(`${queryString} input[type="password"]+i`);
+
+        this.valid = valid;
+        this.disabled = disabled;
+        this.required = required;
```
Comment (tvps20) [inline src/components/password.components.js:10]: I don't get object value
image

Comment (tvps20) [reply src/components/password.components.js:10]: These are states I want, other information can be private.

File: src/components/password.components.js [hunk 17]
```diff
@@ -0,0 +1,97 @@
...
+        if (input.length === 0) {
+            throw new Error('Failed to query input password');
+        }
+
+        if (icon.length === 0) {
+            throw new Error('Failed to query icon');
+        }
+    }
+
+    addClickEvent() {
```
Comment (tvps20) [inline src/components/password.components.js]: I don't have methods for these behaviors
image

File: src/components/password.components.js [hunk 18]
```diff
@@ -0,0 +1,97 @@
...
+                this.icon.removeClass('fa-eye').addClass('fa-eye-slash');
+            }
+        });
+    }
+
+    getValue() {
+        return this.input.val();
+    }
+
+    setValue(string) {
```
Comment (tvps20) [inline src/components/password.components.js]: This comportament is wrong
image

File: src/components/password.components.js [hunk 19]
```diff
@@ -0,0 +1,97 @@
...
+                this.input.attr('type', 'text').trigger('focus');
+                this.icon.removeClass('fa-eye-slash').addClass('fa-eye');
+            } else {
+                this.input.attr('type', 'password').trigger('focus');
+                this.icon.removeClass('fa-eye').addClass('fa-eye-slash');
+            }
+        });
+    }
+
+    getValue() {
```
Comment (tvps20) [inline src/components/password.components.js]: These are types of the comportaments want, other information can be private.
image

**Review by tvps20 — CHANGES_REQUESTED** [review]

File: src/assets/scss/input-password.scss [hunk 20]
```diff
@@ -7,22 +7,21 @@
...
     font-size: 16px;
     font-weight: bold;
-    box-shadow: 2px 2px 4px rgba(0, 0, 0, 0.1);
     margin-bottom: 12px;
     margin-left: 15px;
 }
 
 // Input Password => placeholder => 'Senha'
-.password-input, .password-input:focus {
+.password-input, .password-input:focus, .password-input:disabled {
```
Comment (EvertonDuarteGuimaraes) [inline src/assets/scss/input-password.scss]: Maybe it's more readable if you try something like:
.password {
....&-input, &-input:focus, &-input:disabled {
........ // styles
....}
}
That's take anything with follows password-input:'something'.

**Comment by pedroemanuel42** [comment]
> Pr title is wrong image

Fixed.

File: src/assets/scss/input-password.scss (hunk 20 above)
Comment (pedroemanuel42) [reply src/assets/scss/input-password.scss]: Did it, thanks for the tips!

**Comment by pedroemanuel42** [comment]
> Your description of how has this been tested is confusing. First don't possible how to know test. Must create a new object? Or just run the server?

Fixed.

**Review by feelsantiago — DISMISSED** [review]

**Review by EvertonDuarteGuimaraes — DISMISSED** [review]

File: src/components/password.components.js [hunk 21]
```diff
@@ -0,0 +1,90 @@
...
+        this._input = $(`${queryString} input[type="password"]`);
+        this._icon = $(`${queryString} input[type="password"]+i`);
+
+        this.value = '';
+        this.required = required;
+        this.valid = false;
+        this.disabled = false;
+
+        this._checkQuery(this._div, this._input, this._icon);
+        this.addClickEvent();
```
Comment (tvps20) [inline src/components/password.components.js]: this property can private.

File: src/components/password.components.js [hunk 22]
```diff
@@ -0,0 +1,90 @@
...
+        this._icon.on('click', () => {
+            if (this._input.attr('type') === 'password') {
+                this.showValue();
+            } else {
+                this.hideValue();
+            }
+        });
+    }
+
+    removeClickEvent() {
```
Comment (tvps20) [inline src/components/password.components.js]: why does this behavior exist if disable() already meets the requirements?

**Review by tvps20 — APPROVED** [review]

**Review by tvps20 — CHANGES_REQUESTED** [review]
Leave two questions to analyze

File: src/assets/scss/input-password.scss [hunk 23]
```diff
@@ -3,46 +3,44 @@
...
-    color: $other-dark;
-    background-color: $primary-dark;
-    border: 2px solid $accent-light;
-    border-radius: 20px;
-    height: 50px;
-    box-shadow: 2px 2px 4px rgba(0, 0, 0, 0.1);
-    padding-left: 35px;
-    padding-top: 5px;
-}
+.password- {
```
Comment (feelsantiago) [inline src/assets/scss/input-password.scss]: Remove dash -

File: src/components/tests/password.components.test.js [hunk 24]
```diff
@@ -0,0 +1,242 @@
...
+describe('[PasswordComponents]', () => {
+    let password;
+
+    describe('[_checkQuery]', () => {
+        it('Should throw error if query fail or the div element is not finded.', () => {
+            const div = `<div id="pass"></div>`;
+            $(document.body).html(div);
+
+            expect(() => {
+                password = new PasswordInput('#pazz');
```
Comment (feelsantiago) [inline src/components/tests/password.components.test.js:14]: Don't instantiate global variables in tests. Use beforeEach or BeforeAll methods to setup data. Every test should not affect variables from others tests.

Comment (feelsantiago) [reply src/components/tests/password.components.test.js:14]: Use local variables or resete variables on afterEach or afterAll methods

**Review by feelsantiago — CHANGES_REQUESTED** [review]

**Review by feelsantiago — APPROVED** [review]

**Review by tvps20 — APPROVED** [review]

**Review by tvSantiago — APPROVED** [review]

**Pull request merged**